### PR TITLE
enforce deadlines

### PR DESCRIPTION
**Pull request by HegePI**

### CODE REVIEW

File: services/headless-lms/server/src/controllers/course_material/exercises.rs [hunk 1]
```diff
@@ -96,6 +98,15 @@ async fn post_submission(
     let mut conn = pool.acquire().await?;
     let exercise = models::exercises::get_by_id(&mut conn, *exercise_id).await?;
 
+    // Checking if exercise has a deadline and if so whether it is passed
+    if let Some(deadline) = exercise.deadline {
+        if Utc::now() >= deadline {
```
Comment (nygrenh) [inline services/headless-lms/server/src/controllers/course_material/exercises.rs:103]: 1 second grace time to account clock drift.

Take the deadline by default from chapter and let exercise.deadline to override it.

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/server/src/controllers/course_material/exercises.rs [hunk 2]
```diff
@@ -98,9 +98,23 @@ async fn post_submission(
     let mut conn = pool.acquire().await?;
     let exercise = models::exercises::get_by_id(&mut conn, *exercise_id).await?;
 
+    let chapter = models::chapters::get_chapter(&mut conn, exercise.chapter_id.unwrap()).await?;
```
Comment (nygrenh) [inline services/headless-lms/server/src/controllers/course_material/exercises.rs:101]:
```rust
let course: OptionFuture<_> = exercise.chapter_id.map(|id| models::chapters::get_chapter(&mut conn, id))
.into(); 
```

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/server/src/controllers/course_material/exercises.rs [hunk 3]
```diff
@@ -98,26 +99,35 @@ async fn post_submission(
...
+    let chapter = chapter_option_future.await.transpose()?;
 
-    // Checking if exercise has a deadline and if so whether it is passed
-    if let Some(deadline) = exercise.deadline {
-        // current date added 1 second gracetime to account for clock drift
-        if Utc::now() + Duration::seconds(1) >= deadline {
-            return Err(ControllerError::BadRequest(
-                "exercise deadline passed".to_string(),
-            ));
+    // Exercise deadlines takes precedence to chapter deadlines
```
Comment (nygrenh) [inline services/headless-lms/server/src/controllers/course_material/exercises.rs:109]:
```rust
if let Some(deadline) = exercise.deadline.or(chapter.and_then(|c| c.deadline)) {} 
```

**Review by nygrenh — CHANGES_REQUESTED** [review]

**Review by nygrenh — CHANGES_REQUESTED** [review]
Make the notification smaller and take the new design from figma

File: services/headless-lms/server/src/controllers/course_material/exercises.rs (hunk 3 above)
Comment (HegePI) [reply services/headless-lms/server/src/controllers/course_material/exercises.rs:109]:
```rust
if let Some(deadline) = exercise.deadline.or(chapter.and_then(|c| c.deadline)) {}
```
replaced with
```rust
    if let Some(deadline) = exercise
        .deadline
        .or_else(|| chapter.and_then(|c| c.deadline))
```
clippy suggestion

**Review by nygrenh — CHANGES_REQUESTED** [review]
`${ new Date().toLocaleString("fi", {dateStyle: "long", timeStyle: "short"})} (UTC${-(new Date().getTimezoneOffset() / 60).toString().replace(".", ":")})`

needs to look like this:

 "11. huhtikuuta 2022 klo 14.49 (UTC+03:00)"

```
  const { t, i18n } = useTranslation()
  const lang = i18n.language
```

**Review by nygrenh — APPROVED** [review]

**Pull request merged**